README.md
# stagecoach2

## Stability

Beta. Still experimental, but seeing use in staging environments.

## Configuration

**Step 1.** Create `/usr/local/etc/stagecoach.json` on your server, as follows:

```
{
  "projects": {
    "project-name": {      
      "key": "password-you-create"
    },
  }
}
```

Your `key` should be made up of URL-safe characters.

**Step 2.** As root, install the stagecoach2 npm package globally:

```
npm install -g stagecoach2
```

**Step 3.** As root, make sure that file is readable by the user you'll be running your deployments and your Node.js apps as:

```
chown nodeapps.nodeapps /usr/local/etc/stagecoach.json
```

**Step 4.** Create the `/opt/stagecoach` folder and give it to the appropriate user:

```
mkdir /opt/stagecoach
chown -R nodeapps.nodeapps /opt/stagecoach
```

**Step 4.** While logged in as the appropriate user, or via `su`, tell stagecoach to install itself for automatic restart and to start now:

```
su - nodeapps
stagecoach install
```

> `stagecoach install` will add crontab entries to restart your projects at boot, and to restart stagecoach itself if it stops. You don't have to use this command, if you prefer to set up a `systemd` unit file or other alternative. Such approaches usually require root access.

**Step 5.** Configure your proxy server, such as `nginx`, to direct traffic from the `/stagecoach` subdirectory of your site to port `4000` instead of your application. For simplicity, **do not** rewrite the path part of the URL. Leave it intact.

**Step 6.** Add a github action like this. Note that this action distinguishes between staging and production servers based on the branch name.

```
name: Deploy to Production
on:
  push:
    branches:
      # Save time by only starting up for branches that are deployed somewhere
      - main
      - develop
jobs:
  deploy:
    steps:
    - run: |
        if [ "${GITHUB_REF##*/}" == "main" ]; then
          HOST=mysite.com
        else
          HOST=staging.mysite.com
        fi
        rm -f ./project.tar.gz
        git archive HEAD --format tar | gzip > ./project.tar.gz
        echo "POSTing the project tarball to the server:"
        curl -F 'tarball=@./project.tar.gz' https://${HOST}/stagecoach/deploy/project-name?key=$DEPLOY_KEY
      env:
        DEPLOY_KEY: ${{ secrets.deploy_key }}
```

**Step 7.** Create a **secret** in github (Settings -> Secrets). Name it `deploy_key` and paste the same value you specified for `key` in `stagecoach.json`.

**Step 8.** Push any trivial change to the branch you specified in your github action to verify success. You can see the progress in github actions.

## Edge cases

That's it for typical cases. Here are some common edge cases.

### Ignoring `package-lock.json`

Normally npm respects `package-lock.json`, which keeps you safe from surprises. However if your goal is to **test the bleeding edge** by updating all of your dependencies on deployment, you'll likely need to ignore the `package-lock.json` file when installing dependencies. To do that, you can set the `ignorePackageLock: true` option in `stagecoach.json` as a sub-property of any project.

### For users of stagecoach classic

> This module is unrelated to substack's "stagecoach" npm module, which was published once and never updated 10 years ago. "Stagecoach classic" refers to the [bash-based deployment system](https://github.com/apostrophecms/stagecoach) created at P'unk Avenue and ApostropheCMS.

`stagecoach2` can be used as a drop-in replacement for stagecoach classic, eliminating the need to type `sc-deploy`. You can even continue to use `stagecoach` temporarily while transitioning to git push-based deployment with `stagecoach2`.

Some use cases of stagecoach classic, such as rollbacks or restarts, are not currently covered in the same way by `stagecoach2`. You can roll back with `git revert` and `git push`. There is currently no convenience feature for restarts, but you can restart by triggering a new deployment.

If you are already using stagecoach classic then your root `/opt/stagecoach` folder will probably belong to root. Changing that is not mandatory, but you'll have to create `/opt/stagecoach/locks` and `/opt/stagecoach/logs` and give them to your non-root Node.js user (`nodeapps` in our examples), since `stagecoach` won't be able to create them on its own.

If you use `stagecoach install` to set up stagecoach2, be aware it adds a cron job to start your projects at boot time and be sure to remove any `/etc/rc.d/rc.local` entry running the stagecoach classic `sc-start-all` command in order to avoid conflict.

### Deploying your project when a module is updated

Sometimes projects and npm modules are developed in tandem, or a project is intended as a proving ground for a module. In this situation, it can be helpful to have a test environment where the project depends on the main github branch of the module, and the project is redeployed when *either the module or the project changes.*

In such a case, your Github Action will need to `git clone` the actual project and POST that instead:

```
name: Deploy Project to Production
on:
  push:
    branches:
      # Only pushing to this branch name will deploy
      - main
jobs:
  deploy:
    steps:
    - run: |
        rm -f ./project.tar.gz
        mkdir project
        cd project
        # Public project. For a private project you would need a deployment key
        git clone https://github.com/org/project
        cd project
        git archive HEAD --format tar | gzip > ./project.tar.gz
        echo "POSTing the project tarball to the server:"
        curl -F 'tarball=@./project.tar.gz' https://mysite.com/stagecoach/deploy/project-name?key=$DEPLOY_KEY
```

### Deploying multiple branches of the same project to the same server

For economy you might want to test multiple branches on the same project on the same server.

To do that, configure them as separate projects in `stagecoach.json`, and configure your github actions for each branch to POST to the right project.

Please note however that your project code will not automatically realize that it is separate from other branches on the same server and will compete for access to the same database unless you take steps to distinguish the database name via `__dirname`, which will be different for each deployed branch. With ApostropheCMS, this means setting `shortName` based on the appropriate component of `__dirname`.


CHANGELOG.md
## 0.3

Branch name could contain a `/`, so use a `*` wildcard in the route.

## 0.2

Log file names now incorporate project and branch names in addition to timestamps, preventing a conflict when two deployments kick off simultaneously.

## 0.1

First alpha release of `stagecoach2`, which replaces the bash-based `stagecoach`.


server.js
const passport = require('passport');
const OAuth2Strategy = require('passport-oauth2');
const express = require('express');
const app = express();
const { stripIndent } = require('common-tags');
const session = require('express-session');
const fs = require('fs');
const fetch = require('node-fetch');

const config = JSON.parse(fs.readFileSync(`${__dirname}/config.json`));

app.use(session({
  secret: 'ap9823hrq809273hr0q237y4',
  resave: false, // don't save session if unmodified
  saveUninitialized: false, // don't create session until something stored
  // store: new SQLiteStore({ db: 'sessions.db', dir: './var/db' })
}));

app.use(passport.initialize());
app.use(passport.session());

passport.use(
  new OAuth2Strategy({
    authorizationURL: 'https://github.com/login/oauth/authorize',
    tokenURL: 'https://github.com/login/oauth/access_token',
    clientID: config.githubApp.id,
    clientSecret: config.githubApp.secret,
    callbackURL: "http://localhost:3000/auth/github/callback",
    scope: [ 'read:user' ]
  },
  async (accessToken, refreshToken, _profile, cb) => {
    const response = await fetch('https://api.github.com/user', {
      headers: {
        Authorization: `token ${accessToken}`,
        Accept: 'application/vnd.github+json'
      }
    });
    if (response.status >= 400) {
      return cb(response.status);
    }
    const profile = await response.json();
    return cb(null, profile);
  })
);

passport.serializeUser(function(user, cb) {
  return cb(null, user);
});

passport.deserializeUser(function(user, cb) {
  return cb(null, user);
});

app.get(
  '/auth/github',
  passport.authenticate('oauth2')
);

app.get(
  '/auth/github/callback',
  passport.authenticate('oauth2', {
    failureRedirect: '/?error=login'
  }),
  function(req, res) {
    // Successful authentication, redirect home.
    res.redirect('/');
  }
);

app.get(
  '/',
  (req, res) => {
    return res.send(stripIndent`
<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <title>Home</title> 
  </head>
  <body>
    ${req.user ? stripIndent`
      <p>Hello, ${req.user.login}!</p>
    ` : stripIndent`
      <a href="/auth/github">Log in with Github</a>
    `
    }
  </body>
</html>`
    );
  }
);

app.listen(3000);


app.js
const express = require('express');
const fs = require('fs-extra');
const dayjs = require('dayjs');
const fetch = require('node-fetch');
const util = require('util');
const cp = require('child_process');
const lockfile = require('lockfile');
const quote = require('shell-quote').quote;
const qs = require('qs');
const sluggo = require('sluggo');
const os = require('os');
const lock = util.promisify(lockfile.lock);
const unlock = util.promisify(lockfile.unlock);

const app = express();
// Accept either format that github can be configured to send
app.use(require('body-parser').json({ limit: '10mb' }));
app.use(require('body-parser').urlencoded({ extended: true, limit: '10mb' }));

const argv = require('boring')();
const configFile = process.env.CONFIG || '/usr/local/etc/stagecoach.json';
let config;
let deploying = 0;
let exitAfterDeploy = false;

fs.watch(configFile, { persistent: false }, readConfig);
fs.watch(__filename, { persistent: false }, () => {
  if (deploying > 0) {
    exitAfterDeploy = true;
  } else {
    console.log('Exiting to enable restart with newly installed version of stagecoach');
    process.exit(0);
  }
});

readConfig();

const root = config.root || '/opt/stagecoach';

fs.mkdirpSync(`${root}/logs/deployment`);

app.post('/stagecoach/deploy/:project/*', async (req, res) => {
  const host = req.get('Host');
  if (req.body.payload) {
    // urlencoded option for a github webhook is just JSON encoded
    // inside a "payload" parameter
    try {
      req.body = JSON.parse(req.body.payload);
    } catch (e) {
      if (e) {
        return res.status(400).send('payload POST parameter is not JSON encoded');
      }
    }
  }
  if (!host) {
    return res.status(400).send('missing Host header');
  }
  if (!req.body.ref) {
    return res.send('Thanks but I only care about push events');
  }
  if (!has(config.projects, req.params.project)) {
    return res.status(404).send('no such project');
  }
  const project = config.projects[req.params.project];
  if (!project) {
    return res.status(404).send('no such project');
  }
  project.name = req.params.project;
  if (!req.query.key) {
    return res.status(400).send('missing key query parameter');
  }
  if (req.query.key !== project.key) {
    return res.status(403).send('incorrect key');
  }
  const branchName = req.params[0];
  if (!branchName) {
    return res.status(400).send('missing branch portion of URL');
  }
  if (!has(project.branches, branchName)) {
    return res.send('Thanks but no branch by that name is configured for deployment');
  }
  const branch = {
    ...project.branches[branchName],
    name: branchName
  };
  
  const expectedGithubBranchName = req.query.trigger || branchName;
  if (req.body.ref !== `refs/heads/${expectedGithubBranchName}`) {
    console.log(`ignoring push for ${req.body.ref}, expected ${expectedGithubBranchName}`);
    return res.send(`ignoring push for ${req.body.ref}, expected ${expectedGithubBranchName}`);
  } else {
    console.log(`accepted push for ${req.body.ref}, which matches ${expectedGithubBranchName}`);
  }
  const timestamp = dayjs().format('YYYY-MM-DD-HH-mm-ss');
  const logName = `${sluggo(project.name)}-${sluggo(branchName)}-${timestamp}.log`;
  res.send('deploying');
  // Wait one second before we tell Slack where the logs are, in case they are not ready yet
  setTimeout(function() {
    slack(project, branch, `Starting deployment to ${branch.name}, you may view and refresh logs at https://${host}/stagecoach/logs/deployment/${logName}`);
  }, 1000);
  fs.mkdirpSync(`${root}/locks`);
  try {
    await deploy(project, branch, timestamp, logName);
    slack(project, branch, `👍 Deployment to ${branch.name} SUCCESSFUL, you may view the logs at https://${host}/stagecoach/logs/deployment/${logName}`);
  } catch (e) {
    console.error(e);
    slack(project, branch, `⚠️ Deployment to ${branch.name} FAILED with error code ${e.code || e}, you may view the logs at https://${host}/stagecoach/logs/deployment/${logName}`);
  } finally {
    deploying--;
    if (exitAfterDeploy && (deploying === 0)) {
      console.log('Exiting to enable restart with newly installed version of stagecoach');
      process.exit(0);
    }
  }
});

app.use('/stagecoach/logs', express.static(`${root}/logs`));

app.get('/stagecoach/logs/*', function (req, res) {
  const path = `${root}/logs/${req.params[0]}`;
  if (path.match(/\.\./)) {
    return res.status(400).send('invalid');
  }
  if (!path.match(/\.log$/)) {
    return res.status(404).send('not found');
  }
  return res.send(`
<!DOCTYPE html>
<html>
<head>
  <style>
    #log {
      background-color: black;
      color: #8f8;
      font-family: Monaco, monospace;
      font-size: 16px;
      max-width: 1140px;
      margin: auto;
      overflow: scroll;
      padding: 1em;
    }
  </style>
</head>
<body>
  <pre id="log">
    Connecting...
  </pre>
  <script>
    (() => {
      setTimeout(update, 2500);
      async function update() {
        const url = location.href;
        const log = document.querySelector('#log');
        let finished = false;
        let response;
        try {
          response = await fetch(url + '.txt');
          if (response.status === 404) {
            response = await fetch(url + '.final.txt');
            if (response.status < 400) {
              finished = true;
            }
          }
        } catch (e) {
          console.error(e);
          setTimeout(update, 5000);
          return;
        }
        const text = await response.text();
        let atBottom = ((window.innerHeight + window.scrollY) >= document.body.scrollHeight);
        log.innerText = text;
        if (atBottom) {
          window.scrollTo(0, document.body.scrollHeight);
        }
        if (!finished) {
          setTimeout(update, 2500);
        }
      }
    })();
  </script>
</body>
</html>
`.trim());
});

if (argv._[0] === 'install') {
  install();
} else if (argv._[0] === 'deploy') {
  simulateWebhook();
} else if (argv._[0]) {
  usage();
} else {
  server();
}

function install() {
  console.log('Installing via cron');
  let crontab = '';
  try {
    crontab = cp.execSync('crontab -l', { encoding: 'utf8' });
  } catch (e) {
    if (e.stderr.match(/no crontab/)) {
      console.log('Creating crontab for the first time');
    } else {
      throw e;
    }    
  }
  if (crontab.match(/stagecoach/)) {
    console.log('Already scheduled in cron.');
  } else {
    crontab = crontab.replace(/\n$/, '') + '\n* * * * * stagecoach --if-not-running\n';
    const child = cp.exec('crontab');
    child.on('close', code => process.exit(code));
    child.stdin.write(crontab);
    child.stdin.end();
  }
}

// Send a simulated github webhook to trigger a deployment via the CLI

async function simulateWebhook() {
  const project = config.projects[argv._[1]] || usage();
  const branch = config.projects[argv._[1]].branches[argv._[2]] || usage();
  const response = await fetch(`${branch.baseUrl || project.baseUrl}/stagecoach/deploy/${argv._[1]}/${argv._[2]}?` + qs.stringify({
    key: project.key
  }), {
    method: 'POST',
    headers: {
      'Content-type': 'application/json'
    },
    body: JSON.stringify({
      ref: `refs/heads/${argv._[2]}`
    })
  });
  console.log(await response.text());
  if (response.status >= 400) {
    process.exit(response.status);
  }
}

async function server() {
  // Default port number well out of conflict with typical stagecoach ports
  const port = process.env.PORT || 4000;
  try {
    await listen(port);
    console.log(`Listening on port ${port}`);
  } catch (e) {
    if (argv['if-not-running']) {
      process.exit(0);
    }
    console.error(e);
    process.exit(1);
  }
}

async function deploy(project, branch, timestamp, logName) {
  const lockFile = `${root}/locks/deploy.lock`;
  const logFile = `${root}/logs/deployment/${logName}.txt`;
  const shortName = branch.shortName || project.shortName || project.name;
  const dir = `${root}/apps/${shortName}`;
  await fs.mkdirpSync(dir);
  const checkout = `${dir}/checkout`;
  const current = `${dir}/current`;
  const deployTo = `${dir}/deployments/${timestamp}`;
  let stopped = false;
  let unlinked = false;
  let former;
  let log;
  let locked = false;
  let updated = false;
  deploying++;
  log = await createWriteStream(logFile);
  log.write('Waiting for deployment lock...\n');
  await lock(lockFile, { wait: 60 * 60 * 1000, stale: 59 * 60 * 1000 });
  locked = true;
  log.write('Deployment lock obtained\n');
  let tempScript;
  try {
    const gitSpawnOptions = {
      env: {
        ...process.env
      }
    };
    if (project['ssh-key']) {
      // Support git 1.x, which only has GIT_SSH, not GIT_SSH_COMMAND, so we
      // have to build a temporary script
      tempScript = `${os.homedir()}/${project.name}-connect`;
      fs.writeFileSync(tempScript, `#!/bin/bash
ssh -i ${project['ssh-key']} -o IdentitiesOnly=yes $*`
      );
      fs.chmodSync(tempScript, 0o700);
      gitSpawnOptions.env.GIT_SSH = tempScript;
    }
    const beforeConnecting = existsInCheckout('deployment/before-connecting');
    if (fs.existsSync(checkout)) {
      try {
        if (branch.ignorePackageLock || project.ignorePackageLock) {
          const packageLock = `${checkout}/package-lock.json`;
          if (fs.existsSync(packageLock)) {
            fs.removeSync(packageLock);
          }
        }
        await spawnInCheckout('git', [ 'pull' ], gitSpawnOptions);
        log.write('Deploying commit: ');
        await logCommitId();
        if (beforeConnecting) {
          await spawnInCheckout('npm', [ 'install' ]);
        }
        updated = true;
      } catch (e) {
        log.write('git pull or npm install failed, checking out from scratch: ' + e);
        await fs.remove(checkout);
      }
    }
    if (!updated) {
      await spawn('git', [ 'clone', '--single-branch', '--branch', branch.name, project.repo, checkout ], gitSpawnOptions);
      log.write('Deploying commit: ');
      await logCommitId();
      if (beforeConnecting) {
        await spawnInCheckout('npm', [ 'install' ]);
      }
    }
    if (beforeConnecting) {
      await spawnInCheckout('bash', [ 'deployment/before-connecting' ]);
    }
    const keep = project.keep || 5;
    const deployments = `${dir}/deployments`;
    fs.mkdirpSync(deployments);
    const exclude = existsInCheckout(`deployment/rsync_exclude.txt`) ? '--exclude-from=deployment/rsync_exclude.txt' : '';
    // -C excludes many things related to version control, add back "core" because it is
    // not an uncommon folder name in npm modules
    log.write('syncing to deployment folder...\n');
    await spawnInCheckout('rsync', [ '-C', '-a', '--delete', ...(exclude ? [ '--exclude-from=deployment/rsync_exclude.txt' ] : []), '--include', 'core', '.', deployTo ]);
    if (existsInDeployTo('deployment/dependencies')) {
      // Includes safe migrations
      log.write('Running dependencies script (npm install takes a while)...\n');
      await spawnInDeployTo('bash', [ 'deployment/dependencies' ]);
    }
    let former;
    if (fs.existsSync(current)) {
      try {
        log.write('Stopping old deployment...\n');
        await spawnScriptInCurrent('deployment/stop');
        stopped = true;
      } catch (e) {
        console.warn('🤔 cannot stop current deployment, that may be OK');
      }
      former = fs.readlinkSync(current);
    }
    // Unsafe migrations, if any
    log.write('Running unsafe migrations...\n');
    await spawnInDeployTo('bash', [ 'deployment/migrate' ]);
    console.error(`Removing ${current}`);
    await fs.remove(current);
    unlinked = true;
    log.write('Running start...\n');
    console.log(`|| F: ${deployTo} C: ${current}`);
    await fs.symlink(deployTo, current, 'dir');
    await spawnScriptInCurrent('deployment/start');
    const deploymentsList = fs.readdirSync(deployments).sort();
    if (deploymentsList.length > keep) {
      log.write(`Removing ${deploymentsList.length - keep} older deployments, keeping ${keep}\n`);
      for (let i = 0; (i < deploymentsList.length - keep); i++) {
        const remove = `${deployments}/${deploymentsList[i]}`;
        log.write(`Removing ${remove}\n`);
        await fs.remove(remove);
      }
    }
    log.write('Deployment complete!');
  } catch (e) {
    if (log) {
      log.write('Error on deployment:\n');
      log.write(e + '\n');
    }
    console.error(e);
    if (unlinked) {
      log.write('Relinking previous deployment\n');
      await fs.remove(current);
      console.log(`<< F: ${former} C: ${current}`);
      await fs.symlink(former, current, 'dir');
    }
    await fs.remove(deployTo);
    if (stopped) {
      await spawnScriptInCurrent('deployment/start');
    }
    throw e;
  } finally {
    if (log) {
      await log.close();
      await fs.rename(logFile, logFile.replace('.txt', '.final.txt'));
    }
    if (tempScript) {
      if (fs.unlinkSync(tempScript)) {
        fs.unlinkSync(tempScript);
      }
    }
    if (locked) {
      await unlock(lockFile);
    }
  }

  // awaitable
  function spawn(cmd, args = [], options = {}) {
    options = {
      ...{
        stdio: [ 'pipe', log, log ]
      },
      ...options
    };
    const child = cp.spawn(cmd, args, options);
    return new Promise((resolve, reject) => {
      child.on('close', code => {
        if (code) {
          return reject(code);
        } else {
          return resolve(null);
        }
      });
      child.on('error', e => reject(e));
    });
  }

  async function spawnInCheckout(cmd, args = [], options = {}) {
    options = {
      ...{
        cwd: checkout
      },
      ...options
    };
    return spawn(cmd, args, options);
  }

  function existsInCheckout(path) {
    return fs.existsSync(`${checkout}/${path}`);
  }

  async function spawnInDeployTo(cmd, args = [], options = {}) {
    options = {
      ...{
        cwd: deployTo
      },
      ...options
    };
    return spawn(cmd, args, options);
  }

  function existsInDeployTo(path) {
    return fs.existsSync(`${deployTo}/${path}`);
  }

  // Run a specific bash shell script, with no arguments, with the current
  // working directory set to ${project}/current, but
  // without resolving cwd to an absolute path. This is useful to scripts
  // like legacy forever-based "stop" and "start" scripts that want
  // the path name to be part of a stable forever id, even though the
  // target of "current" changes
  async function spawnScriptInCurrent(script) {
    // We can't use the cwd option of spawn because node always resolves
    // it to an absolute path, so do it another way
    return spawn('bash', [ '-c', `(cd ${quote([current])} && bash ${quote([script])})` ], {});
  }

  function existsInCurrent(path) {
    return fs.existsSync(`${current}/${path}`);
  }

  async function logCommitId() {
    await spawnInCheckout('git', [ 'rev-parse', 'HEAD' ]);
  }

}

function slack(project, branch, text) {
  const webhook = branch.slackWebhook || project.slackWebhook || config.slackWebhook;
  if (webhook) {
    return fetch(
      webhook, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({
        text
      })
    });
  }
}

function has(o, k) {
  return Object.hasOwnProperty.call(o, k);
}

// Once awaited the stream is really ready and can be passed to stdio of spawn
function createWriteStream(path) {
  const stream = fs.createWriteStream(path);
  return new Promise((resolve, reject) => {
    stream.on('open', () => resolve(stream));
  });
}

async function listen(port) {
  const server = app.listen(port);
  return new Promise((resolve, reject) => {
    server.on('error', reject);
    server.on('listening', resolve);
  });
}

function readConfig() {
  config = JSON.parse(fs.readFileSync(configFile, 'utf8')); 
}

function usage() {
  console.error(`
Usage: stagecoach [command] [arguments]

With no command specified, this command will listen for connections as a stagecoach
deployment server.

"stagecoach install" will install a cron job to ensure such a deployment server
is running at all times. It will start up within one minute after installing
the cron job.

"stagecoach deploy projectName branchName" will trigger a deployment by sending
a simulated github push webhook. This is mainly for testing.
  `.trim());
  process.exit(1);
}
